File, Search Blocks: Lower CSS specificity

MaggieCabrera opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

Lowers the CSS specificity for the File and Search Block selectors

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

The high specificity of these selectors are overriding the newest addition to the elements API (the buttons), pointed out [here](https://github.com/WordPress/gutenberg/pull/40260#issuecomment-1138289624) and [here](https://github.com/WordPress/gutenberg/pull/41239#issuecomment-1138409634).

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

I'm trying to remove unneeded selectors and used :where when I wasn't sure about what I could remove without breaking other stuff

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->

I've tested this using emptytheme, with the following markup:

```
<!-- wp:buttons -->
<div class="wp-block-buttons"><!-- wp:button -->
<div class="wp-block-button"><a class="wp-block-button__link wp-element-button">Fill Button</a></div>
<!-- /wp:button -->

<!-- wp:button {"className":"is-style-outline"} -->
<div class="wp-block-button is-style-outline"><a class="wp-block-button__link wp-element-button">Outline Button</a></div>
<!-- /wp:button --></div>
<!-- /wp:buttons -->

<!-- wp:search {"label":"Search","buttonText":"Search"} /-->

<!-- wp:search {"label":"Search","buttonText":"Search","buttonUseIcon":true} /-->

<!-- wp:search {"label":"Search","buttonText":"Search","buttonPosition":"button-inside"} /-->

<!-- wp:search {"label":"Search","buttonText":"Search","buttonPosition":"button-inside","buttonUseIcon":true} /-->

<!-- wp:file {"id":23,"href":"http://maggie.local/wp-content/uploads/2022/05/Screenshot-2022-05-13-at-13.10.27.png"} -->
<div class="wp-block-file"><a id="wp-block-file--media-4fff6e6c-0ba4-4897-9486-9c2bfc9e36f6" href="http://maggie.local/wp-content/uploads/2022/05/Screenshot-2022-05-13-at-13.10.27.png">Screenshot 2022-05-13 at 13.10.27</a><a href="http://maggie.local/wp-content/uploads/2022/05/Screenshot-2022-05-13-at-13.10.27.png" class="wp-block-file__button wp-element-button" download aria-describedby="wp-block-file--media-4fff6e6c-0ba4-4897-9486-9c2bfc9e36f6">Download</a></div>
<!-- /wp:file -->
```

And in the theme.json file I have this:

```
"styles": {
      "elements": {
	      "button": {
		      "border": {
			      "radius": "4px"
		      },
		      "color": {
			      "background": "blue",
			      "text": "yellow"
		      },
		      "typography": {
			      "fontSize": "22px",
			      "fontWeight": "strong",
			      "lineHeight": "2",
			      "textDecoration": "none"
		      },
		      "spacing": {
			      "padding": {
				      "top": "2em",
				      "bottom": "2em",
				      "left": "4em",
				      "right": "4em"
			      }
		      }
	      }
      }
 }
```

## Screenshots or screencast <!-- if applicable -->

With the theme.json changes:

Trunk | This PR
--- | ---
<img width="970" alt="Screenshot 2022-05-27 at 11 03 51" src="https://user-images.githubusercontent.com/3593343/170672907-101fd6e2-3589-41bb-b488-3b27ca189c8f.png"> | <img width="931" alt="Screenshot 2022-05-27 at 11 26 51" src="https://user-images.githubusercontent.com/3593343/170672913-45d2d312-577e-4638-bc17-90082a75e4b6.png">


Without the theme.json changes:


Trunk | This PR
--- | ---
<img width="937" alt="Screenshot 2022-05-27 at 11 32 51" src="https://user-images.githubusercontent.com/3593343/170673069-92722bd4-2ef0-4ea7-b857-7b5cfac8a15b.png"> | <img width="1002" alt="Screenshot 2022-05-27 at 11 32 37" src="https://user-images.githubusercontent.com/3593343/170673065-0e6c8f94-b706-4cdc-abc4-4779d9657615.png">


/cc @WordPress/block-themers

MaggieCabrera committed:
lower specificity on file block

MaggieCabrera committed:
lower specificity for search block

draganescu commented on packages/block-library/src/search/style.scss:
```diff
@@ -47,7 +47,7 @@
 }
 
-.wp-block-search.wp-block-search__button-inside .wp-block-search__inside-wrapper {
+:where(.wp-block-search__button-inside) :where(.wp-block-search__inside-wrapper) {
```
What does this rule do?

MaggieCabrera replied:
What do you mean? what does the `:where` do?

draganescu replied:
No 😂 I mean the two together. I understand `:where(x)` but `:where(x) :where(x)` is unclear. It's a bit of an [unexpected formula](https://developer.mozilla.org/en-US/docs/Web/CSS/:where).

MaggieCabrera replied:
mmh, from what I understand `:where(.class) :where(.class)` is the same as `:where(.class .class)`, so I don't have a preference

MaggieCabrera replied:
maybe the second is more readable

draganescu replied:
Yes the second one is more readable. To me `:where(.class) :where(.class)` reads like ` * ` :)))

MaggieCabrera committed:
made selector more readable, fixed specificity of nested element

mikachan reviewed (approved):
This works for me when using emptytheme and those theme.json settings 👍 I think the use of `:where` works well here too.

draganescu reviewed (changes requested):
🧇 great! @MaggieCabrera can we add a comment for 

`* + .wp-block-file__button` and `:where(.wp-block-search__button-inside .wp-block-search__inside-wrapper)`  explaining how we want to lower specificity and for what reason? Given someone may come back to this months later it's easy to no understand what's up with this solution.

Similar to how we do [here](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/paragraph/style.scss#L40) in the `paragraph` block.

MaggieCabrera committed:
make a comment on the where rule

MaggieCabrera commented:
I´m not changing `* + .wp-block-file__button`, just moving `.wp-block-file__button` outside of `.wp-block-file`, but the diff is picking it as if I changed all the code. I added the comment on the other change though, to make it clearer

MaggieCabrera committed:
add
ed another comment

draganescu reviewed (approved):
LGTM 🚢

sabernhardt commented:
Could the Search block's _wrapper_ have a higher specificity?

As noted on [Trac 60164](https://core.trac.wordpress.org/ticket/60164), the old CSS reset removed border and padding on any `div` element, which would override `:where(.wp-block-search__button-inside .wp-block-search__inside-wrapper)` in many themes.

Using two elements could be enough:
```
:where(.wp-block-search__button-inside .wp-block-search__inside-wrapper) {
	// a little more specific than one element selector for wrapper styles
	html body & {
		padding: $grid-unit-05;
		border: 1px solid $gray-600;
		box-sizing: border-box;
	}
}
```
